gobblin-runtime/src/main/java/gobblin/runtime/MultiConverter.java

```diff
@@ -80,87 +79,122 @@ public Object convertSchema(Object inputSchema, WorkUnitState workUnit)
+      throw new UnsupportedOperationException();
+    }
+
+    /**
+     * A helper class that implements {@link Iterator}. It is constructed with a {@link Iterator} and a {@link Converter}.
+     * The class iterates through the results of each converted record from prevIterator. It iterates through each
+     * element in prevIterator, converts the result, and then stores the result in the currentIterator object. It
+     * returns every element in currentIterator until it is empty, and then it gets the next element from prevIterator,
+     * converts the object, and stores the result in currentIterator. This pattern continues until there are no more
+     * elements left in prevIterator.
+     */
+    private class ChainedConverterIterator implements Iterator<Object> {
+
+      private final Converter converter;
+      private final Iterator<Object> prevIterator;
+
+      private Iterator<Object> currentIterator;
+
+      public ChainedConverterIterator(Iterator<Object> prevIterator, Converter converter)
+          throws DataConversionException {
+        this.converter = converter;
+        this.prevIterator = prevIterator;
+
+        if (this.prevIterator.hasNext()) {
+          this.currentIterator =
+              converter.convertRecord(convertedSchemaMap.get(converter), this.prevIterator.next(), workUnitState)
+                  .iterator();
+        } else {
+          this.currentIterator = new EmptyIterable<Object>().iterator();
+        }
       }
 
-      if (this.converterIteratorStack.isEmpty()) {
+      @Override
+      public boolean hasNext() {
+        if (this.currentIterator.hasNext()) {
+          return true;
+        }
+        while (this.prevIterator.hasNext()) {
+          try {
+            this.currentIterator =
+                converter.convertRecord(convertedSchemaMap.get(converter), this.prevIterator.next(), workUnitState)
+                    .iterator();
+          } catch (DataConversionException e) {
+            Throwables.propagate(e);
+          }
+          if (this.currentIterator.hasNext()) {
+            return true;
+          }
+        }
         return false;
       }
 
-      // Get the next record from the iterator (that still has more records) at the top of the stack
-      // and reconstruct the stack of converter iterators from that record
-      for (int i = this.converterIteratorStack.size(); i < converters.size(); i++) {
-        Object record = this.converterIteratorStack.peek().next();
-        Converter converter = converters.get(i);
-        try {
-          Iterator<Object> iterator =
-              converter.convertRecord(convertedSchemaMap.get(converter), record, this.workUnitState).iterator();
-          this.converterIteratorStack.push(iterator);
-          // Do not continue when encountering an empty iterator
-          if (!iterator.hasNext()) {
-            break;
+      @Override
+      public Object next() {
+        if (this.currentIterator.hasNext()) {
```

Comment:
You can replace the duplicate code here with a call to `hasNext()`, and throw a `NoSuchElementException` if `hasNext` returns false.